Codechange: Remove StationIDStack and SmallStack.

PeterN opened the pull request:
<!--
Commit message:

- Please use Feature / Add / Change / Fix for player-facing changes. E.g.: "Feature: My cool new feature".
- Please use Feature / Add / Change / Fix followed by "[NewGRF]" or "[Script]" for moddable changes. E.g.: "Feature: [NewGRF] My cool new NewGRF addition".
- Please use Codechange / Codefix for developer-facing changes. E.g.: "Codefix #1234: Validate against nullptr properly".

See https://github.com/OpenTTD/OpenTTD/blob/master/CODINGSTYLE.md#commit-message for more details.
-->

## Motivation / Problem

`SmallStack` is convoluted code, concerned with memory management and avoiding copies, and uses a pool to store items (not a regular Pool any more, though it was once.). It comes from a time before we embraced stl containers.

The only user is `StationIDStack`, and because of shallow copies it's difficult to see what's going on.

It seems to be used to build a list of `StationID`s, and then pop from the back in various ways.

<!--
Describe here shortly
* For bug fixes:
    * What problem does this solve?
    * If there is already an issue, link the issue, otherwise describe the problem here.
* For features or gameplay changes:
    * What was the motivation to develop this feature?
    * Does this address any problem with the gameplay or interface?
    * Which group of players do you think would enjoy this feature?
-->


## Description

Use a `std::vector` or `std::span` instead. This more clearly delineates when the list of `StationID`s is owned, and is simpler to understand.

<!--
Describe here shortly
* For bug fixes:
    * How is the problem solved?
* For features or gameplay changes:
    * What does this feature do?
    * How does it improve/solve the situation described under 'motivation'.
-->


## Limitations

I may well have misunderstood what the existing implementation does and how it behaves.

We don't have unit tests to verify the behaviour.

<!--
Describe here
* Is the problem solved in all scenarios?
* Is this feature complete? Are there things that could be added in the future?
* Are there things that are intentionally left out?
* Do you know of a bug or corner case that does not work?
-->


## Checklist for review

Some things are not automated, and forgotten often. This list is a reminder for the reviewers.
* The bug fix is important enough to be backported? (label: 'backport requested')
* This PR touches english.txt or translations? Check the [guidelines](https://github.com/OpenTTD/OpenTTD/blob/master/docs/eints.md)
* This PR affects the GS/AI API? (label 'needs review: Script API')
    * ai_changelog.hpp, game_changelog.hpp need updating.
    * The compatibility wrappers (compat_*.nut) need updating.
* This PR affects the NewGRF API? (label 'needs review: NewGRF')
    * newgrf_debug_data.h may need updating.
    * [PR must be added to API tracker](https://wiki.openttd.org/en/Development/NewGRF/Specification%20Status)

PeterN committed:
Codechange: Remove StationIDStack and SmallStack.

Use a std::vector or std::span instead.

rubidium42 commented on src/order_cmd.cpp:
```diff
@@ -392,9 +392,9 @@ StationIDStack OrderList::GetNextStoppingStation(const Vehicle *v, VehicleOrderI
 			} else {
-				StationIDStack st1 = this->GetNextStoppingStation(v, skip_to, hops);
-				StationIDStack st2 = this->GetNextStoppingStation(v, advance, hops);
-				while (!st2.IsEmpty()) st1.Push(st2.Pop());
```
I haven't checked or tested this, but doesn't this line add the StationIDs from the end to the begin? At least I'd assume `Pop()` on a stack returns the last element. Especially since the `rbegin()/rend()` I have seen earlier.

PeterN replied:
Yes, it is now done in reverse. I am ignoring the question of whether it actually matters, and whether this is still the correct way to collect all destinations given the change from stack to vector.

PeterN replied:
I think the second branch was returned in reverse purely because that's the only way the stack worked, not for any particular need. Keeping that behaviour means instantiating vectors, while it may not even be required.

So the additional new commit removes the overhead of allocating more vectors by having the caller pass one vector that the function keeps adding to. I'm not sure under what conditions a sub-branch can actually return multiple destinations, so it's possible that for most cases this doesn't even change behaviour.

rubidium42 reviewed (commented):
It definitely makes the code easier to understand as we don't need to understand the weird magic in the small stack code.

PeterN committed:
Codechange: Use single vector for next station order.

Pass vector to insert into to avoid handling multiple vectors.

This may change the order of returned stations.
